### ISSUE TITLE
[dcl.struct.bind] structured bindings vs. potentially conflict

### ISSUE THREAD

**Issue by xmh0511** [cplusplus/draft#5201]
As have discussed in this thread: https://stackoverflow.com/questions/65729849/it-seems-the-current-standard-draft-cannot-interpret-why-two-structured-binding  

````cpp
struct A{
    int a;
};
struct B{
    int b;
};
auto&& [x] = A{};  //#1
auto&& [x] = B{};  //#2
int main(){
}
````
There is no provision in the current draft that can point out the name `x` at `#1` potentially conflicts with `x` at `#2`. In short, [basic.scope.scope] p4 is saying they correspond, [basic.link] p8 is saying they declare the **same entity**, obviously, their shared name denotes the same entity, and also [basic.link] p11 didn't even mention **structured bindings**. Finally, [basic.def.odr] also didn't mention it. Furthermore, except for this example, there is also a similar example: 
````cpp
struct C{
    int a; 
    int b;
};
auto [x, x] = C{0,0};
````

[dcl.struct.bind] p1 is saying 
> A structured binding declaration introduces the identifiers v<sub>0</sub>, v<sub>1</sub>, v<sub>2</sub>, ... of the identifier-list as names of structured bindings. 

So, each name of the identifier denotes the corresponding structured binding introduced by a structured binding declaration.  Obviously, [basic.pre] p5 is wrong in this case, which says
> An entity E is denoted by the name (if any) that is introduced by **a** declaration of E or by a typedef-name introduced by a declaration specifying E.  

I would argue that a structured binding declaration can introduce multiple structured binding**s**. Obviously, the relationship between them is not one-to-one.  Moreover, [basic.pre] p3 admits that "structured binding" itself is a distinct category of the entity. So, we should regulate which structured bindings declare the same entity or not.

**Comment by jensmaurer** [comment]
A /declaration/ can have several declarations inside, and I think that's exactly what a structure binding declaration does: `[a, b] = f()` declares two structured bindings (entities), a and b.

@opensdh, it looks like [basic.link] p11 should mention structured bindings.

basic.def.odr, after my rewrite in a recent core issue, should already say the right thing that we can't have a redefinition of a structured binding.

**Comment by xmh0511** [comment]
@jensmaurer 
> A /declaration/ can have several declarations inside  

Sure, `using A  = struct B{};` is an instance. However, a structured binding declaration is a bit different since there are no more declarations nested in the structured binding declaration.   

> An entity E is denoted by the name (if any) that is introduced by a declaration of E

The wording "a declaration of E" sounds like that the declaration is uniquely possessed by the entity `E`. This makes the following example hard to interpret
````cpp
auto [a, b] = f(); // #1
int c, d;  // #2
`````
Either at `#1` or `#2`, there is only **one declaration**, which can introduce more than one entity. the entities denoted by `a` and `b` **share** the declaration at `#1`, `c` and `d` share a declaration at `#2`. Either declaration is **not** uniquely possessed by one of the entities. So, I think we may change the above sentence to that    

> An entity E is denoted by the name (if any) that is introduced by a declaration introducing E.  

which can eliminate the puzzle of the "ownership".

**Comment by jensmaurer** [comment]
As I said, the model here is that `int a, b;` is two declarations (prose term) (one for a, one for b), even though it is only one (grammatical) _declaration_. With that interpretation, `auto [x,x] = f()` is clearly ill-formed. (The declaration of a structured binding is a definition, and we have duplicate definitions for "x" here.)

**Comment by xmh0511** [comment]
> As I said, the model here is that int a, b; is two declarations (prose term) (one for a, one for b)

I agree with this point since we have [dcl.decl.general] p3 that explicitly clarifies this point  

> Each init-declarator or member-declarator in **a declaration** is analyzed **separately as if it were in a declaration by itself**.

In other words, each init-declarator is considered to be in a single corresponding declaration.  

But, how about a structured binding declaration? we never define a similar concept to clarify that each structured binding has its own declaration for analysis.

**Comment by jensmaurer** [comment]
> But, how about a structured binding declaration?

It should be treated the same, but we seem to be lacking words for that effect.

**Comment by xmh0511** [comment]
> but we seem to be lacking words to that effect.

Well, that's a part of what I'm looking forward to in this issue.  We should specify that part in the current draft.
